daffodil-runtime1/src/main/scala/org/apache/daffodil/layers/LayerTransformer.scala

```diff
@@ -0,0 +1,272 @@
+    maybeLayerTerminatorEv: Maybe[LayerTerminatorEv],
+    trd: TermRuntimeData): LayerTransformer
+}
+
+/**
+ * Transformers have factories. This lets you find the transformer factory
+ * by the name obtained from dfdl:layerTransform.
+ */
+object LayerTransformerFactory {
+
+  private lazy val transformerMap = new NonAllocatingMap(new HashMap[String, LayerTransformerFactory])
+
+  def register(factory: LayerTransformerFactory): Unit = {
+    transformerMap.put(factory.name, factory)
+  }
+  /**
+   * Given name, finds the factory for the transformer. SDE otherwise.
+   *
+   * The state is passed in order to provide diagnostic context if not found.
+   */
+  def find(name: String, state: ParseOrUnparseState): LayerTransformerFactory = {
+    val maybeFactory = transformerMap.get(name)
+    if (maybeFactory.isEmpty) {
+      val choices = transformerMap.keySet.mkString(", ")
+      state.SDE("The dfdl:layerTransform '%s' was not found. Available choices are: %s", name, choices)
+    } else {
+      maybeFactory.get
+    }
+  }
+
+  /**
+   * All transformers must be registered so they are available by name.
+   *
+   * It is possible to package a transformer in a separate jar also, but then
+   * someone has to register it by calling the register method.
+   *
+   * Transformers built into the primary Daffodil jars/packages should be
+   * registered here.
+   */
+  register(Base64MIMETransformerFactory)
+}
+
+/**
+ * Shared functionality of all LayerTransformers.
+ */
+abstract class LayerTransformer()
+  extends Serializable {
+
+  protected def wrapLayerDecoder(jis: java.io.InputStream): java.io.InputStream
+
+  def wrapJavaInputStream(s: DataInputStream, fInfo: FormatInfo): java.io.InputStream = {
+    new JavaIOInputStream(s, fInfo)
+  }
+
+  /**
+   *  Override these if we ever have transformers that don't have these
+   *  requirements.
+   */
+  val mandatoryLayerAlignmentInBits: Int = 8
+  val mandatoryLengthUnit: LayerLengthUnits = LayerLengthUnits.Bytes
+
+  def addLayer(s: DataInputStream, fInfo: FormatInfo): DataInputStream = {
+    val jis = wrapJavaInputStream(s, fInfo)
+    val decodedInputStream = wrapLayerDecoder(jis)
+    val newDIS = ByteBufferDataInputStream(decodedInputStream, 0L)
+    newDIS.cst.setPriorBitOrder(BitOrder.MostSignificantBitFirst) // must initialize priorBitOrder
+    newDIS
+  }
+
+  def removeLayer(s: DataInputStream): Unit = {
+    // nothing for now
+  }
+
+  def addLayer(s: DataOutputStream, fInfo: FormatInfo): DataOutputStream = ???
+
+  def removeLayer(s: DataOutputStream): Unit = ???
+
+}
+
+class JavaIOInputStream(s: DataInputStream, finfo: FormatInfo)
+  extends java.io.InputStream {
+
+  private lazy val id = Misc.getNameFromClass(this)
+
+  private val readBuf = new Array[Byte](1)
+
+  override def read(): Int = {
+    val n = read(readBuf, 0, 1)
+    if (n != -1)
+      readBuf(0)
+    else
+      n
+  }
+
+  override def read(b: Array[Byte]) =
+    read(b, 0, b.length)
+
+  /**
```

Comment:
Revise this javadoc. This was copied here for reference when writing the code, but really this is scala code and this needs to be scaladoc, or just omitted since this is a well-documented interface already.

Reply:
By saying "this is scala code", is this copied from the scala library. If so, we need to make sure it's the LICENSE file is updated, and ideally this would be in a scalaLib directory like some other scala stuff.